polly/src/event_manager.rs

```diff
@@ -0,0 +1,786 @@
+        None
+    }
+    /// Handle an error event (EPOLLERR).assert_ne!
+    fn handle_error(&mut self, _source: Pollable) -> Option<Vec<PollableOp>> {
+        None
+    }
+
+    /// Initial registration of pollable objects.
+    /// Use the PollableOpBuilder to build the vector of PollableOps.
+    fn init(&self) -> Option<Vec<PollableOp>>;
+}
+
+/// Wraps a HashMap of fd to EventHandlerData.
+struct HandlerMap {
+    handlers: HashMap<i32, EventHandlerData>,
+}
+
+impl HandlerMap {
+    pub fn new() -> HandlerMap {
+        HandlerMap {
+            handlers: HashMap::new(),
+        }
+    }
+
+    pub fn get_mut_ref(&mut self, id: i32) -> Option<&mut EventHandlerData> {
+        self.handlers.get_mut(&id)
+    }
+
+    // Returns a copy of EventHandlerData instead of mut ref.
+    pub fn get(&mut self, id: i32) -> Option<EventHandlerData> {
+        match self.handlers.get(&id) {
+            Some(handler) => Some(EventHandlerData::new(
+                (handler.data.0.clone(), handler.data.1),
+                handler.handler.clone(),
+            )),
+            None => None,
+        }
+    }
+}
+
+impl Deref for HandlerMap {
+    type Target = HashMap<i32, EventHandlerData>;
+    fn deref(&self) -> &Self::Target {
+        &self.handlers
+    }
+}
+
+impl DerefMut for HandlerMap {
+    fn deref_mut(&mut self) -> &mut HashMap<i32, EventHandlerData> {
+        &mut self.handlers
+    }
+}
+
+// The tx side of a generic API channel.
+// Sending a T message over this channel will also write the eventFd.
+// The rx side of the channel polls the eventfd for incoming messages.
+pub struct GenericApiChannel<T> {
+    channel: Sender<T>,
+    fd: EventFd,
+}
+
+impl<T> GenericApiChannel<T> {
+    pub fn new(channel: Sender<T>) -> GenericApiChannel<T> {
+        GenericApiChannel {
+            channel,
+            fd: EventFd::new().unwrap(),
+        }
+    }
+
+    /// Send a message of type T and notify rx side by writing
+    /// the eventfd.
+    pub fn send(&mut self, msg: T) -> Result<()> {
+        self.fd.write(1).map_err(Error::ApiChannelFd)?;
+        self.channel.send(msg).map_err(|_| Error::ApiChannelTx)
+    }
+
+    /// Reads evenfd event count.
+    pub fn read_event(&mut self) -> u64 {
+        self.fd.read().unwrap_or(0)
+    }
+}
+
+impl<T> Clone for GenericApiChannel<T> {
+    fn clone(&self) -> Self {
+        GenericApiChannel {
+            channel: self.channel.clone(),
+            fd: self.fd.try_clone().unwrap(),
+        }
+    }
+}
+
+impl<T> AsRawFd for GenericApiChannel<T> {
+    fn as_raw_fd(&self) -> RawFd {
+        self.fd.as_raw_fd()
+    }
+}
+
+pub type ApiChannelMessage = (WrappedHandler, Vec<PollableOp>);
+pub type ApiChannel = GenericApiChannel<ApiChannelMessage>;
+
+pub struct EventManager {
+    fd: Pollable,
+    handlers: HandlerMap,
+    events: Vec<epoll::Event>,
+    api_channel_rx: Receiver<ApiChannelMessage>,
+    api_channel_tx: ApiChannel,
```

Comment:
Why do we need to explicitly declare the api_channel_rx and tx here?

Reply:
The rx side is managed by the EventManager in process_api_channel() and the tx side clones are exposed to EventHandler implementors via get_api_channel().

Reply:
Polly seems to be an independent crate. From this point of view I wasn't expecting to see objects that are related to the way our Firecracker VMM works and it looks like the API is just one of these objects.

I was wondering if this can be avoided in any way and if it makes sense, but I think it's better to sync offline.

Reply:
After our discussion I understand that api_channel_* are not actually related to the Firecracker API server, but just gives access to the communication channel. Can we rename the variable to make it less confusing?